File: app/src/main/java/cn/ifafu/ifafu/electricity/http/CookieInterceptor.java
package cn.ifafu.ifafu.electricity.http;

import androidx.annotation.NonNull;

import java.io.IOException;

import cn.ifafu.ifafu.app.Constant;
import cn.ifafu.ifafu.electricity.util.CookieUtils;
import cn.ifafu.ifafu.util.SPUtils;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

public class CookieInterceptor implements Interceptor {

    @NonNull
    @Override
    public Response intercept(@NonNull Chain chain) throws IOException {
        Request.Builder builder = chain.request().newBuilder();
        String cookie = CookieUtils.getCookie();
        if (!cookie.isEmpty()) {
            builder.addHeader("Cookie", cookie);
        }
        Response response = chain.proceed(builder.build());
        if (!response.headers("Set-Cookie").isEmpty()) {
            for (String header: response.headers("Set-Cookie")) {
                String[] kv = header.substring(0, header.indexOf(";")).split("=");
                SPUtils.get(Constant.SP_ELEC).putString(kv[0], kv[1]);
            }
        }
        return response;
    }
}

File: app/src/main/java/cn/ifafu/ifafu/data/retrofit/interceptor/CookieInterceptor.kt
package cn.ifafu.ifafu.data.retrofit.interceptor

import cn.ifafu.ifafu.data.repository.RepositoryImpl
import cn.ifafu.ifafu.data.entity.ElecCookie
import okhttp3.Interceptor
import okhttp3.Response
import java.io.IOException

class CookieInterceptor : Interceptor {

    private var cookie: ElecCookie? = null

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val builder = chain.request().newBuilder()
        if (cookie == null) {
            cookie = RepositoryImpl.XfbRt.getElecCookie()
        }
        builder.addHeader("Cookie", cookie!!.toCookieString())
        val response = chain.proceed(builder.build())
        if (response.headers("Set-Cookie").isNotEmpty()) {
            for (header in response.headers("Set-Cookie")) {
                println(header)
                val kv = header.substring(0, header.indexOf(";")).split("=").toTypedArray()
                println("Cookie put: " + kv.contentToString())
                if (kv.size < 2) {
                    cookie!![kv[0]] = ""
                } else {
                    cookie!![kv[0]] = kv[1]
                }
            }
            RepositoryImpl.XfbRt.saveElecCookie(cookie!!)
        }
        return response
    }
}